### DISCUSSION TITLE
Alternative Executor to replace CeleryExecutor

### DISCUSSION THREAD

**Discussion by hadarsharon** [apache/airflow#41902] [Q&A]
Hi everyone,

I work with an Airflow deployment on AWS EKS that leverages both CeleryExecutor and KubernetesExecutor, and while KubernetesExecutor seems to scale nicely and seldom has any issues, the CeleryExecutor has been giving us headaches pretty much ever since we started scaling heavily (we have over 100k tasks daily). 
Recently we've also done work to integrate AWS' ECS Executor into the mix, now that hybrid executors are a thing. 

Now I'm not really insinuating anything about Celery's ability to scale, but I'm interested in exploring whether I have any other options other than Celery because it seems to be hard to maintain and configure without the occasional hiccups and random errors (workers receving warm shutdowns, random SIGTERMs with no apparent cause, etcetera). 

In the past there was a DaskExecutor that was since deprecated, and IIRC, a MesosExecutor. I've even seen something about a RayExecutor (probably a community provider that never took off) - could anyone suggest anything other than sticking to KubernetesExecutor? Cheers

**Comment by o-nikolas** [comment]
Hey @hadarsharon

Agreed that Celery can be a bit unstable at large scales. The community is still working to improve this (here's [a PR](https://github.com/apache/airflow/pull/41870) in flight as we speak for example!) that should improve scaling if you're using the DB as your results backend.

> Recently we've also done work to integrate AWS' ECS Executor into the mix, now that hybrid executors are a thing.

Love to see this, I was involved in the creation of both the ECS executor and the multiple executor configuration feature. Glad to see both being used! :) 

> Now I'm not really insinuating anything about Celery's ability to scale, but I'm interested in exploring whether I have any other options other than Celery

I'm curious to hear what aspects of Celery you like and want to keep (this will help me nudge towards the best executor for your usecase). But if you're looking for another executor that supports a queue which workers read tasks from and workers can run more than one task then the [AwsBatchExecutor](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/executors/batch-executor.html) does exactly those things. It supports scaling the worker pool up and down really nicely too (as well as keeping a baseline number of workers running and ready to receive tasks immediately) and easy controls for how many tasks run on each worker.

**Reply by hadarsharon** [reply]
> that should improve scaling if you're using the DB as your results backend.

I am using the DB as the results backend, yes. Would it be better to have a separate DB for it that isn't Airflow's metadata DB, or does it not matter?

FWIW, Redis is the broker (we've tried RabbitMQ in the past but it didn't fare any better). 

> Love to see this, I was involved in the creation of both the ECS executor and the multiple executor configuration feature. Glad to see both being used! :)

They are both great additions to Airflow, kudos to you! :) Definitely been waiting for this to happen. 

> I'm curious to hear what aspects of Celery you like and want to keep

Sure, I will elaborate a bit:

1. To get the obvious out of the way - I like that it can run an arbitrary amount of tasks on a single worker, and the cluster bit with autoscaling via KEDA. 
2. When it works - it works great, quite snappy and responsive to the scheduler, autoscaling is a breeze, workers come and go and it allows us to run many more tasks than we could if we'd be using LocalExecutor or KubernetesExecutor solely. 
3. The feature I would most appreciate is for a cluster manager/scheduler/"master" if you will, to be able to load-balance the tasks across workers not just by round-robin but by actually examining which worker is the best one to accept the task. I'm not too familiar with Celery's internals and its intricacies, but I'm not sure it actually does that? Still, when there is more than one worker up, the load seems to be balanced alright. 
4. I like the **idea** of Celery Flower - a dashboard where you can track the cluster is obviously a great tool, but the implementation is not so good imo. I've had more positive experience profiling and examining Dask's dashboard, for reference. 

My main issues with Celery, in the context of using it as an Executor in Airflow are:

- The dreaded `Error/Reason: Executor CeleryExecutor(parallelism=1024) reported that the task instance <TaskInstance: some_dag.some_task [queued]> finished with state failed, but the task instance's state attribute is queued.` - I still see it mentioned in issues every now and then, and I still don't know where exactly the issue lies and whether or not anyone's managed to fix it. It seems to happen almost exclusively with Celery, whereas with KubernetesExecutor we really almost never get that. 
- Flower being almost useless (mentioned earlier)
- Airflow doesn't have enough (any?) pointers as to how to manage Celery workers in the context of dedicated workers for certain workloads, and I'll explain: let's say I want to dedicate 2 out of my 4 celery workers only for HTTP-based tasks that store API responses in memory. I think Celery itself allows for one to spin up a worker that listens only to a certain queue, but I haven't found any documentation in Airflow as to how to control that via the Helm chart, let alone with autoscaling. 
This means that whenever a worker spins up, it shares whatever tasks you throw at it and is in a valid queue for Celery workers. That creates some issues in case 1 out of, say, 16 tasks suddenly decides to load something too large in memory and the whole worker gets SIGTERMed. 

---

As for AwsBatchExecutor - I did come across it, but it seemed more geared towards heavy-resource, longer tasks such as ML training and the like - not sure I'd wanna run something like a 5 second SQL query on that? 

The closest equiavlent I could think of to CeleryExecutor would be DaskExecutor, but it seems that it's been pretty much neglected. I would happily start maintaining it and try to push some life back into it so as to be a legitimate executor in Airflow, but I'm not sure how difficult that would be and whether Airflow would even want to support that in any way. Thoughts?

Thank you for your comment :)

**Reply by o-nikolas** [reply]
> As for AwsBatchExecutor - I did come across it, but it seemed more geared towards heavy-resource, longer tasks such as ML training and the like - not sure I'd wanna run something like a 5 second SQL query on that?

I'm not sure where that notion came from, if you found it somewhere in our docs due let me know! Because that's not the case at all. It services the exact same usecases as Celery. You can use EC2 hosts of any size to have a fleet of autoscaled hosts or even use ECS fargate tasks for quick tasks. From what you've described it fits all your usecases. It also has the AWS console, which allows you to monitor and observe your workloads, which in a way is Flower but actually useful.

Regarding assigning tasks to different workers, you can use queues to attache workers of specific types to specific queues and then send your tasks to those specific job queues. Both Celery and Batch support this and would address that usecase.

**Reply by hadarsharon** [reply]
> I'm not sure where that notion came from, if you found it somewhere in our docs due let me know! Because that's not the case at all. It services the exact same usecases as Celery. You can use EC2 hosts of any size to have a fleet of autoscaled hosts or even use ECS fargate tasks for quick tasks. From what you've described it fits all your usecases. It also has the AWS console, which allows you to monitor and observe your workloads, which in a way is Flower but actually useful.

I meant that the start times are (allegedly) somewhat slow (see [Best Practices](https://docs.aws.amazon.com/batch/latest/userguide/best-practices.html#bestpractice1) on AWS docs) whereas currently with Celery it picks up the task almost instantly. 
Perhaps that was an old limitation/quirk of Batch that's been patched since and now it's almost instantaneous, but if we take Fargate as an example, you can consider the 50-60 seconds worth of start time whenever you launch a task - that's gotta suck for running a large number of SQL queries concurrently and sequentially, each taking anywhere between 1 second to 10 seconds to finish. In Celery it's not so much an issue as it would be in Fargate and, if the rumours are true - Batch. 

I am willing to give it a try of course, but the idea will probably be scrapped if that is the case :/

> Regarding assigning tasks to different workers, you can use queues to attache workers of specific types to specific queues and then send your tasks to those specific job queues. Both Celery and Batch support this and would address that usecase.

Sure; However, when using Airflow on Kubernetes with KEDA for autoscaling, is there any native support in Airflow for scaling up a worker attached to queue X based on a number of tasks being sent to queue X, and at the same time doing the same thing for queues Y and Z? I haven't noticed that anywhere in the Helm chart, it seems to be naïvely checking for any tasks in the default queue (and not `kubernetes` if using CeleryKubernetesExecutor) and then spinning up a generic Celery worker. 

I imagine that I could modify the Celery worker launch to attach to a specific queue, but I don't see how I can make KEDA scale workers dynamically on a per-queue basis directly via the Airflow Helm chart without sprinkling some Kustomizations (K intended) or thereabout - am I wrong?

**Reply by o-nikolas** [reply]
> I meant that the start times are (allegedly) somewhat slow (see [Best Practices](https://docs.aws.amazon.com/batch/latest/userguide/best-practices.html#bestpractice1) on AWS docs) whereas currently with Celery it picks up the task almost instantly.
Perhaps that was an old limitation/quirk of Batch that's been patched since and now it's almost instantaneous, but if we take Fargate as an example, you can consider the 50-60 seconds worth of start time whenever you launch a task - that's gotta suck for running a large number of SQL queries concurrently and sequentially, each taking anywhere between 1 second to 10 seconds to finish. In Celery it's not so much an issue as it would be in Fargate and, if the rumours are true - Batch.

> I am willing to give it a try of course, but the idea will probably be scrapped if that is the case :/

This is just a common confusion. Batch is almost too flexible and hard to understand. Batch can support multiple compute backends. If the compute you're backing Batch with is ephemeral (e.g. ECS Fargate) then yes you pay the startup cost when batch needs to launch your container. But Batch can also work exactly like Celery where you can keep a fleet of workers running (EC2 for sure and I believe ECS containers also) and ready to accept tasks immediately as they arrive in the queue just like Celery and then autoscale them up and down as required as well as how many tasks run per worker, etc.


>  Sure; However, when using Airflow on Kubernetes with KEDA for autoscaling, is there any native support in Airflow for scaling up a worker attached to queue X based on a number of tasks being sent to queue X, and at the same time doing the same thing for queues Y and Z? I haven't noticed that anywhere in the Helm chart, it seems to be naïvely checking for any tasks in the default queue (and not kubernetes if using CeleryKubernetesExecutor) and then spinning up a generic Celery worker.

I'm not sure about using KEDA or Helm, those aren't my areas of knowledge, but you can certainly do this with Batch. You simply attach autoscaling groups to each job queue and then batch will scale that group up (given the parameters and limits you've provided) when it receives jobs to that queue.
